Disallow implicit Any types from Silent Imports

ilinum opened the pull request:
These types can appear from an unanalyzed module.
If mypy encounters a type annotation that uses such a type,
it will report an error.

Note that this PR is different from #3141 although the flags have similar names. Perhaps, we should figure out the proper names for all flags.

JukkaL commented on test-data/unit/check-flags.test:
```diff
@@ -376,3 +376,52 @@ if y:  # E: Condition must be a boolean
+  return l
+[builtins fixtures/list.pyi]
+[out]
+main:5: error: Return type is implicitly converted to 'builtins.list[Any]' due to import from unanalyzed module
```
Would it make sense to update the types to be formatted as `List[Any]` etc. for consistency with most other messages? `builtins.list[Any]` and such are pretty verbose and not how users write the types.

JukkaL commented on test-data/unit/check-flags.test:
```diff
@@ -376,3 +376,52 @@ if y:  # E: Condition must be a boolean
+class C(Unchecked): # E: Subclassing type 'Unchecked' that is implicitly converted to 'Any' due to import from unanalyzed module
+  pass
+
+class A(List[Unchecked]): # E: Subclassing a type that is implicitly converted to 'builtins.list[Any]' due to import from unanalyzed module
```
Similar to above (`List[Any]` preferred over `builtins.list[Any]`).

JukkaL commented on test-data/unit/check-flags.test:
```diff
@@ -376,3 +376,52 @@ if y:  # E: Condition must be a boolean
+  pass
+
+class A(List[Unchecked]): # E: Subclassing a type that is implicitly converted to 'builtins.list[Any]' due to import from unanalyzed module
+  pass
```
Ideas for additional test cases:

* Type alias like `X = List[Unchecked]` or `X = Unchecked`.
* `cast(List[Unchecked], foo)`
* `NamedTuple` uses an implicit `Any` type.
* `T = TypeVar('T', Unchecked, str)`
* `T = TypeVar('T', bound=Unchecked)`
* `X = NewType('X', Unchecked)`
* Implicitly converted type within a callable type or a tuple type.

ilinum commented:
The code review feedback should be addressed now. 
But we need to decide on a better name for this flag before this PR is ready to be merged.

ilevkivskyi commented:
FWIW, I think the name of the flag you propose is too generic, e.g. ``List`` is annotation is also implicitly converted to ``List[Any]`` per PEP 484, but this is not prohibited be this flag. Maybe a better name would be ``--disallow-imported-any``?

ilinum commented:
@ilevkivskyi Yes, I defintely agree with you. In https://github.com/python/mypy/issues/3427 I mentioned `--disallow-any-from-unanalyzed-module` but that's a little more verbose than your option.

ddfisher commented on mypy/typeanal.py:
```diff
@@ -232,7 +232,9 @@ def visit_unbound_type(self, t: UnboundType) -> Type:
                     # is pretty minor.
-                    return AnyType()
+                    any_type = AnyType()
+                    any_type.is_from_silent_import = sym.node.type.is_from_silent_import
```
This suggests `is_from_silent_import` to me that should be an optional argument to `AnyType`'s constructor.

ilinum replied:
Yes, I was thinking about it. Would probably make it much cleaner. 

What stopped me was that there is already an parameter `implicit` in the constructor from [this commit](https://github.com/python/mypy/commit/6de11e10b3af3ec0b974a7f69664d63f337003ee).
It should be fine if I comment it though.

ddfisher commented on mypy/typeanal.py:
```diff
@@ -731,6 +733,23 @@ def visit_callable_type(self, t: CallableType) -> TypeVarList:
             return []
 
 
+def has_any_from_silent_import(t: Type) -> bool:
```
Rather than special-casing `Callable` at some of the call sites, it'd be better to handle it here.

ilinum replied:
Responded to this in this discussion: https://github.com/python/mypy/pull/3405#discussion_r118377736

ddfisher commented on mypy/options.py:
```diff
@@ -158,3 +160,6 @@ def module_matches_pattern(self, module: str, pattern: Pattern[str]) -> bool:
     def select_options_affecting_cache(self) -> Mapping[str, bool]:
         return {opt: getattr(self, opt) for opt in self.OPTIONS_AFFECTING_CACHE}
+
+    def silent_mode(self) -> bool:
```
Though actually I think one of the call sites will probably be removed, making this no longer necessary.

ilinum commented on mypy/semanal.py:
```diff
@@ -1428,7 +1438,12 @@ def add_unknown_symbol(self, name: str, context: Context, is_import: bool = Fals
+        any_type = AnyType()
+        if self.options.silent_mode():
+            # if silent mode is not enabled, no need to report implicit conversion to Any,
+            # let mypy report an import error.
```
Yeah, this makes sense. It's not obvious that the new flag should be used with the silent imports.

ilinum commented on mypy/semanal.py:
```diff
@@ -964,7 +965,9 @@ def analyze_base_classes(self, defn: ClassDef) -> None:
-                if self.options.disallow_subclassing_any:
+                # if --disallow-implicit-any-types is set, the issue is reported later
+                implicit = self.options.disallow_implicit_any_types and base.is_from_silent_import
+                if self.options.disallow_subclassing_any and not implicit:
```
Yeah, I think it's okay. As long as it's not overwhelming for the user, it's actually better to have more information than less. The new message would explain more. This is what it outputs now for a simple case:
```
error: Cannot find module named '
note: (Perhaps setting MYPYPATH or using the "--ignore-missing-imports" flag would help)
error: Class cannot subclass 'Unchecked' (has type 'Any')
error: Base type Unchecked is implicitly converted to "Any" due to import from unanalyzed module
```

ilinum commented on mypy/checkexpr.py:
```diff
@@ -1543,8 +1544,11 @@ def visit_cast_expr(self, expr: CastExpr) -> Type:
+        if options.warn_redundant_casts and is_same_type(source_type, target_type):
             self.msg.redundant_cast(target_type, expr)
+        if options.disallow_implicit_any_types and has_any_from_silent_import(target_type):
+            self.msg.implicit_any_from_silent_import("Target type of cast", target_type, expr)
```
I'm not totally sure what you mean. I tried the following code and it gave an error (like it's supposed to).
```
from typing import cast, Callable
from missing import Unchecked

foo = [1, 2, 3]
call = cast(Callable[[], Unchecked], foo)
```
Produces the following output:
```
m.py:6: error: Target type of cast is implicitly converted to Callable[[], Any] due to import from unanalyzed module
```

ilinum replied:
Oh, I see what you're saying. Maybe it should say stuff like `Return type is Any instead of Unchecked` (not exact wording). Is that what you mean?

I am actually not sure about this. In another case we are typechecking a function definition (where the user types in `def foo ...`), so there is an error specific to return type and arguments. However, when the user types in Callable, I think talking about Callable as it's own type makes sense. Do you have a strong opinion on this?

ddfisher replied:
Ah, that was my misunderstanding.  I saw you dissecting `Callable` types [here](https://github.com/python/mypy/pull/3405/files#diff-90cb1d334445b0300466f7da165ae183R615) and assumed that was because `has_any_from_silent_import` didn't properly handle them.  Taking a closer look, you're just trying to give people a better error message.  That seems reasonable!

ilinum commented:
Thanks for the feedback, David! 

Yeah, I realized I commited typeshed after I submitted the PR. It should be good now! Thanks for the advice. 

I responded to all your comments - and will follow it in the future. Thanks!

For now I addressed all of the feedback, except for the `Callable` type.
And we still need to discuss good name for the flag and how to refer to it in comments/variable names.

ddfisher commented on mypy/checker.py:
```diff
@@ -1179,6 +1190,14 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
+        if (s.type is not None and
+                self.options.disallow_implicit_any_types and
+                has_any_from_silent_import(s.type)):
+            if isinstance(s.lvalues[-1], TupleExpr):  # is multiple
```
Nitpick: this comment should be a bit more explicit.  In general, I'd err on the side of trying to be very very clear in comments, even if that feels a little verbose.  In this case, I'd consider moving the comment to the next line and writing something like:
```
# This is a multiple assignment. Instead of figuring out which type is problematic, give a generic error message.
```

ilinum replied:
Yep, that makes sense! Agree.

ddfisher commented on mypy/types.py:
```diff
@@ -272,9 +272,12 @@ def serialize(self) -> JsonDict:
 
-    def __init__(self, implicit: bool = False, line: int = -1, column: int = -1) -> None:
+    def __init__(self, implicit: bool = False, from_silent_import: bool = False,
+                 line: int = -1, column: int = -1) -> None:
```
Nitpick: if you have to wrap a function def line, put each argument on a line of its own.  There's a lot of existing code in mypy which doesn't do this, but it's much easier to read if done this way, so we're trying to do this for new code.

ilinum replied:
Seems reasonable!

ddfisher commented on mypy/types.py:
```diff
@@ -272,9 +272,12 @@ def serialize(self) -> JsonDict:
+                 line: int = -1, column: int = -1) -> None:
         super().__init__(line, column)
-        self.implicit = implicit
+        self.implicit = implicit  # Was this Any type was inferred without a type annotation?
```
Nitpick: probably better to just put this comment on the preceding line for consistency with the next member variable initalization.

ilinum replied:
Yup!

ddfisher commented on mypy/types.py:
```diff
@@ -272,9 +272,12 @@ def serialize(self) -> JsonDict:
         super().__init__(line, column)
-        self.implicit = implicit
+        self.implicit = implicit  # Was this Any type was inferred without a type annotation?
+        # Does this come from an unresolved import? See--disallow-implicit-any-types flag
```
Trivial: `See--` -> `See --`

ilinum replied:
Didn't notice when I was committing! :)

ddfisher commented on test-data/unit/check-flags.test:
```diff
@@ -376,3 +376,133 @@ if y:  # E: Condition must be a boolean
+def f(x: X) -> None:  # E: Argument 1 to "f" is implicitly converted to List[Any] due to import from unanalyzed module
+    pass
+[builtins fixtures/list.pyi]
+[case testDisallowImplicitAnyCast]
```
Trivial: missing blank line before this line.  It's a lot harder to read the tests if there isn't a blank line before each new testcase.

ilinum replied:
Yes! Updated my other tests too.

ilinum commented:
I updated the commit to use the new option name and syntax (for more, see #3470). 

In the near future, this option will accept a list of parameters but for now it only accepts one (`unimported`). After this PR is merged, I will rename `--disallow-untyped-defs` to `--disallow-any=untyped`.

ilinum commented on mypy/messages.py:
```diff
@@ -855,6 +855,11 @@ def unsupported_type_type(self, item: Type, context: Context) -> None:
         self.note('Redundant cast to {}'.format(self.format(typ)), context)
 
+    def unimported_type_becomes_any(self, prefix: str, typ: Type, ctx: Context) -> None:
+        self.fail("{} becomes {} due to an unfollowed import (such imports occur either "
```
I'm not sure it is good to have an error message that long. 

We could remove the explanation in parens but I'm not sure that most users will understand what an `unfollowed import` is otherwise.

@ddfisher what do you think?

ddfisher replied:
I agree -- this error message is too long.  I also agree that it'll probably be somewhat confusing.  I'd say: remove the parenthetical explanation for now, and we can consider some other mechanism of explaining to the user if we think it's a big enough problem.  (But we can/should do that in another PR.)  We could consider outputting one `Note:` at the end if there are any of these errors explaining what an unfollowed import is.

ilinum replied:
Yeah, a note sounds like a better option!

ddfisher commented on mypy/main.py:
```diff
@@ -203,6 +204,16 @@ def add_invertible_flag(flag: str,
+        current_options = raw_options.split(',')
+        for option in current_options:
+            if option not in valid_disallow_any_options:
+                formatted_opts = ', '.join(map("'{}'".format, valid_disallow_any_options))
```
Nit: strongly prefer list comprehensions to map (or filter) in Python.  I'm also a fan of functional programming, but list comprehensions are considered more pythonic.

ddfisher replied:
I'd also consider calling this "formatted_valid_options" in order to disambiguate from the other "*_options" variables.  I'd also write out "options" fully here -- it's a bit more verbose, but is more consistent with the rest of the variable names.

ilinum replied:
Yep!

The reason I chose `map` in that situation was that it was shorter than using a comprehension. 
But it's probably better to use whatever is more pythonic :)

ddfisher replied:
If you're interested, talk to Guido about his views on `map` vs list comprehensions.  (Hint: they are very strong views.)

ilinum replied:
I will, thanks :)

ddfisher commented on mypy/main.py:
```diff
@@ -203,6 +204,16 @@ def add_invertible_flag(flag: str,
             strict_flag_assignments.append((dest, not default))
 
+    def disallow_any_argument_type(raw_options: str) -> List[str]:
+        current_options = raw_options.split(',')
```
I'm not sure this is named quite right.  "current_options" implies that the options will be changing over time, which doesn't really happen.  Maybe consider just "options"?

ilinum replied:
Good point. I was trying to make it clear that these options are the ones provided by user. I think just `options` works too!

ddfisher commented on mypy/main.py:
```diff
@@ -179,6 +179,7 @@ def process_options(args: List[str],
 
     strict_flag_names = []  # type: List[str]
     strict_flag_assignments = []  # type: List[Tuple[str, bool]]
+    valid_disallow_any_options = {'unimported'}
```
Tiny nit: I think the "valid_" part of this name doesn't really add anything (and is very slightly misleading because it isn't only used in a checking-for-validity context: it's also displayed to the user).

ddfisher replied:
More importantly, this should be a list instead of a set, because it's displayed to our users in a couple places.  This is important for two reasons:  1) We don't want a nondeterministic `--help` output.  2) We're going to want to specify the options in a particular order that we think makes the most sense to read them in.

A set is canonically the "right" data structure to use here, because you're mainly checking for membership.  However, it's so small that it doesn't matter (especially because it's not going to be run in a tight loop or anything).

ilinum replied:
Yep, all that makes sense!

Yeah, I didn't think about the fact that the help output could be in different order -- that's a good point!

ddfisher commented on mypy/main.py:
```diff
@@ -203,6 +204,16 @@ def add_invertible_flag(flag: str,
+        for option in current_options:
+            if option not in valid_disallow_any_options:
+                formatted_opts = ', '.join(map("'{}'".format, valid_disallow_any_options))
+                message = "Unrecognized option '{}' (valid options are: {}).".format(
```
This line needs to mention that the option is related to `--disallow-any`.  Much less importantly, I'd also consider "Invalid" instead of "Unrecognized", as it seems slightly more accurate.

I'd consider something like `"Invalid '--disallow-any' option ..."`.

ilinum replied:
Oh wow, yeah good catch. I assumed when the error would be thrown, it would say the option name but I guess not.

Yep!

ddfisher replied:
It's always good to run these things on the command-line yourself to make sure the output really looks as expected!  (And sometimes the way you feel about phrasing, etc, can change when you see it in context.)

ddfisher commented on mypy/main.py:
```diff
@@ -205,14 +205,15 @@ def add_invertible_flag(flag: str,
+        for option in options:
+            if option not in disallow_any_options:
+                formatted_valid_options = ', '.join(
+                    "'{}'".format(option) for option in disallow_any_options)
```
`option` here shadows `option` defined on line 209 -- you really don't want to do that.  I think a one-letter `o` would be fine as a name here (because the context you need to see what that means is all right on the same line).

ilinum replied:
Hmm, good point.
Actually, variable `options` on line 208 overshadows `options` from outer scope (line 372). 
Will fix!

ilinum commented:
Should be good now!

ddfisher commented:
Thanks!
